### ISSUE TITLE
How to debug custom loss function?

### ISSUE THREAD

**Issue by dankal444** [tensorflow/tensorflow#54234]
I know about 2 things needed to debug tensorflow in eager mode:
- add `run_eagerly=True` when compiling model
- add `tf.config.run_functions_eagerly(True)` line

Neither is working when it comes to debugging custom loss function. I used [simple MNIST example](https://www.tensorflow.org/datasets/keras_example) to recreate the issue:

```
import tensorflow as tf
import tensorflow_datasets as tfds

tf.config.run_functions_eagerly(True)  # does not help

(ds_train, ds_test), ds_info = tfds.load(
    'mnist',
    split=['train', 'test'],
    shuffle_files=True,
    as_supervised=True,
    with_info=True,
)

def normalize_img(image, label):
  """Normalizes images: `uint8` -> `float32`."""
  return tf.cast(image, tf.float32) / 255., label


def custom_loss(y_true, y_pred):
    stop = 1   # BREAKPOINT HERE NOT WORKING!


ds_train = ds_train.map(
    normalize_img, num_parallel_calls=tf.data.AUTOTUNE)
ds_train = ds_train.cache()
ds_train = ds_train.shuffle(ds_info.splits['train'].num_examples)
ds_train = ds_train.batch(128)
ds_train = ds_train.prefetch(tf.data.AUTOTUNE)

ds_test = ds_test.map(
    normalize_img, num_parallel_calls=tf.data.AUTOTUNE)
ds_test = ds_test.batch(128)
ds_test = ds_test.cache()
ds_test = ds_test.prefetch(tf.data.AUTOTUNE)

model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dense(10)
])
model.compile(
    optimizer=tf.keras.optimizers.Adam(0.001),
    # loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
    loss=custom_loss,
    metrics=[tf.keras.metrics.SparseCategoricalAccuracy()],
    run_eagerly=True,       # does not help either
)

model.fit(
    ds_train,
    epochs=6,
    validation_data=ds_test,
)
```

**Comment by mohantym** [comment]
Hi @dankal444 ! 2.x versions run eagerly by default and loss function has to be a differential one.  Above code was working after replacing custom loss with this code snippet. You can know more about  creating custom loss from[ here](https://keras.io/api/losses/). Thank you!
```
def custom_loss(y_true, y_pred):
    y_true = tf.cast(y_true,tf.float32)
    y_pred = tf.cast(y_pred,tf.float32)
    squared_difference = tf.square(y_true - y_pred)
    return tf.reduce_mean(squared_difference, axis=-1)  # Note the `axis=-1`
```

**Comment by dankal444** [comment]
Hi @mohantym this is not a question about this specific *toy example*. As you can see custom loss computation is empty, intentionally. I just want the debugger to stop there so it much easier to implement the loss.

**Comment by mohantym** [comment]
@dankal444 ! I have tried to catch the bug using try catch block and tf.print() . Attaching updated [gist](https://colab.sandbox.google.com/gist/mohantym/f21ad71a2cb6bdb19dd4f031b82bd347/github_54234.ipynb#scrollTo=liEiG7qIsARM) for reference. Thanks!

**Comment by dankal444** [comment]
@mohantym What is this gist for? What it has to do with debugging I issue here?

**Comment by mohantym** [comment]
@dankal444 ! I thought your query was on debugging custom loss function. I understand you wanted  to stop at loss function instead of throwing exception.  Please post on  [TF Forum](https://discuss.tensorflow.org/) on further assistance. Thank you!

**Comment by dankal444** [comment]
@mohantym 

> I thought your query was on debugging custom loss function

yes

> I understand you wanted to stop at loss function instead of throwing exception

yes

> I updated gist accordingly to debug loss function

but if you set breakpoint in the loss function it does not work. As I have wrote "I just want the debugger to stop there so it is much easier to implement the loss, try some things, fix shapes, etc."

**Comment by jvishnuvardhan** [comment]
@dankal444 When you run `model.fit`, it creates a graph and I think TF/keras inspects it. In the case where there is nothing returning from a custom loss function (which is must for a model and should be differentiable), it is throwing an error. I am not exactly sure how to make the breakpoint approach to work but this is not a technically a bug. Thanks!

[Here](https://colab.research.google.com/gist/jvishnuvardhan/2ea6433e97aded319007916a716359f1/github_54234.ipynb) is a gist for our reference. If you want to contribute, please feel free to submit any PR to update the code. Thanks!

**Comment by dankal444** [comment]
So, TLDR: you can't..

But there is a workaround - overwriting `train_step` function:
https://keras.io/guides/customizing_what_happens_in_fit/
You can debug there.

Note that not everything that works in eager mode will work in no-eager mode. Had to struggle with many not-supported things
